app/models/patient.rb

```diff
@@ -931,4 +935,132 @@ def new_submission_token
     end
     token
   end
+
+  def handle_update
+    monitoring_change if monitoring_changed?
+    isolation_change if isolation_changed?
+    case_status_change if case_status_changed?
+    symptom_onset_change if symptom_onset_changed?
+    continuous_exposure_change if continuous_exposure_changed?
+  end
+
+  def monitoring_change
+    return if monitoring
+
+    self.closed_at = DateTime.now
+    self.continuous_exposure = false
+  end
+
+  def isolation_change
+    return if isolation
+
+    self.extended_isolation = nil
+    # NOTE: The below will overwrite any new value they may set for symptom onset as they can not be set in the exposure workflow.
+    self.user_defined_symptom_onset = false
+    self.symptom_onset = calculated_symptom_onset
+  end
+
+  def case_status_change
+    return unless ['Suspect', 'Unknown', 'Not a Case'].include?(case_status) && public_health_action != 'None'
+
+    self.public_health_action = 'None'
+  end
+
+  def symptom_onset_change
+    return unless symptom_onset.nil?
+
+    self.user_defined_symptom_onset = false
+    self.symptom_onset = calculated_symptom_onset
+  end
+
+  def continuous_exposure_change
+    self.continuous_exposure = false if continuous_exposure && !monitoring
+  end
+
+  def monitoring_history_edit(history_data, diff_state)
+    patient_before = history_data[:patient_before]
+    history_data[:updates]&.keys&.each do |attribute|
+      updated_value = self[attribute]
+      next if patient_before[attribute] == updated_value
+
+      case attribute
+      when :monitoring
+        History.monitoring_status(history_data)
+
+        # If the record was in continuous exposure and then it was closed and continuous exposure was turned off
+        if !updated_value && patient_before[:continuous_exposure] && !continuous_exposure
+          History.monitoring_change(
+            patient: self,
+            created_by: 'Sara Alert System',
+            comment: 'System turned off Continuous Exposure because the record was moved to the closed line list.'
+          )
+        end
+      when :exposure_risk_assessment
+        History.exposure_risk_assessment(history_data)
+      when :monitoring_plan
+        History.monitoring_plan(history_data)
+      when :public_health_action
+        History.public_health_action(history_data)
+      when :assigned_user
+        History.assigned_user(history_data)
+      when :pause_notifications
+        History.pause_notifications(history_data)
+      when :last_date_of_exposure
+        History.last_date_of_exposure(history_data)
+      when :extended_isolation
+        History.extended_isolation(history_data)
+      when :continuous_exposure
+        History.continuous_exposure(history_data)
+      when :isolation
+        update_patient_history_for_isolation(patient_before, updated_value)
+      when :symptom_onset
+        History.symptom_onset(history_data)
+      when :case_status
```

Comment:
I noticed when testing that because of the order of this switch statement the order of the history messages is now changed and it can be a bit more confusing. 

Before:
![Screen Shot 2021-01-19 at 1 14 11 PM](https://user-images.githubusercontent.com/11698457/105076318-e75a2a80-5a58-11eb-84e7-9cb6a670c46f.png)


After:
![Screen Shot 2021-01-19 at 1 12 40 PM](https://user-images.githubusercontent.com/11698457/105076304-e32e0d00-5a58-11eb-94fb-cd0e5a0066e6.png)

Before, the change that caused the other changes (changing the case status) came first, and now it comes last. Can we move that up in the switch statement to preserve that order?

Reply:
Good call. Fixed in https://github.com/SaraAlert/SaraAlert/pull/664/commits/e63ebf8217e37501ccb885efabca37b2d6148fb3. Since the switch statement is within a loop, and the order that the attributes go through that loop is not guaranteed (aka we don't know if `case_status` or `isolation` will come first), I moved the history update for `isolation` to a separate block of code so that it will always come last.

Reply:
Hmm, I'm thinking we need a better way of handling this.

`extended_isolation` also changes when `isolation` changes (so unfortunately it's not just `case_status` and `isolation`) - so while right now it seems to be working with this change, this is probably also something we'd want to guarantee. I also don't think it's just this case - `monitoring` changes affect `continuous` exposure, `case_status` changes also affect `public_health_action`. 

This is actually a bit of a difficult problem with our current setup. Because users can only change workflows via a case status change on the UI, but that is NOT the case via API so what triggers that kind of change is not consistent between the two. Only thing I can think to do is to do explicit checks within the case blocks for case status, for example, and check to see if isolation has also been changed, and if so handle the history for that directly there (and same for other cases). It wouldn't really be a problem if we could guarantee the order of the keys in `history_data[:updates]` - perhaps we should just sort this?

Reply:
Ok, sorted in https://github.com/SaraAlert/SaraAlert/pull/664/commits/1408283c208b2d10634ae0d25a0c1e3506d986dd. 

The sorting does only handle `case_status` and `isolation` right now, because those are the only two attributes I'm aware of where this could be a problem.

In the case of `monitoring` changes impacting `continuous_exposure`, we shouldn't have any issues, because those History updates are both created in one block of code:
```ruby
when :monitoring
        History.monitoring_status(history_data)

        # If the record was in continuous exposure and then it was closed and continuous exposure was turned off
        if !updated_value && patient_before[:continuous_exposure] && !continuous_exposure
          History.monitoring_change(
            patient: self,
            created_by: 'Sara Alert System',
            comment: 'System turned off Continuous Exposure because the record was moved to the closed line list.'
          )
        end
```
So the update detailing the change in `monitoring_status` will always come before the "Sara Alert System" message about `continuous_exposure`. Same goes for `case_status` affecting `public_health_action`. The situation with `case_status` and `isolation` is unique, because the front end handles the logic to set `isolation` to `false` when `case_status` changes. In all other cases, we do those sort of side effects on the back end. Since the front end handles the change, we get updates for `case_status` and `isolation` side by side, without it being clear that `case_status` caused the change in `isolation`, but the sorting you suggested handles that now.

Reply:
Yeah you're right. The sorting solution is less performant but I think it'll help us if more cases like this pop-up, plus N is small here. Thanks!